sky/data/storage.py

```diff
@@ -2021,3 +2075,392 @@ def _delete_r2_bucket(self, bucket_name: str) -> bool:
+                                    'data transfers between COS and r2')
+                else:
+                    self.batch_ibm_rsync([self.source])
+
+        except Exception as e:
+            raise exceptions.StorageUploadError(
+                f'Upload failed for store {self.name}') from e
+
+    def delete(self) -> None:
+        self._delete_cos_bucket()
+        logger.info(f'{colorama.Fore.GREEN}Deleted COS bucket {self.name}.'
+                    f'{colorama.Style.RESET_ALL}')
+
+    def get_handle(self) -> StorageHandle:
+        return self.s3_resource.Bucket(self.name)
+
+    def batch_ibm_rsync(self,
+                        source_path_list: List[Path],
+                        create_dirs: bool = False) -> None:
+        """Invokes rclone copy to batch upload a list of local paths to cos
+
+        Since rclone does not support batch operations, we construct
+        multiple commands to be run in parallel.
+
+        Args:
+            source_path_list: List of paths to local files or directories
+            create_dirs: If the local_path is a directory and this is set to
+                False, the contents of the directory are directly uploaded to
+                root of the bucket. If the local_path is a directory and this is
+                set to True, the directory is created in the bucket root and
+                contents are uploaded to it.
+        """
+
+        def get_sync_command(src_dir_path, dest_dir_name) -> str:
+            """returns an rclone command that copies files in
+              'from src_dir_path' to 'dest_dir_name'.
+
+            `rclone copy` copies files from source path to target.
+            files with identical names at won't be copied over, unless
+            their modification date is more recent.
+            works similarly to `aws sync` (without --delete).
+
+            Args:
+                src_dir_path (str): source path from which to copy files.
+                dest_dir_name (str): target path where files are copied to.
+
+            Returns:
+                str: bash command using rcolne to sync files
+            """
+            # .git directory is excluded from the sync
+            # wrapping src_dir_path with "" to support sources
+            # containing spaces.
+            sync_command = (
+                'rclone copy --exclude ".git/*" '
+                f'"{src_dir_path}" '
+                f'{self.bucket_rclone_profile}:{self.name}/{dest_dir_name}')
+            return sync_command
+
+        # Generate message for upload
+        if len(source_path_list) > 1:
+            source_message = f'{len(source_path_list)} paths'
+        else:
+            source_message = source_path_list[0]
+
+        with log_utils.safe_rich_status(
+                f'[bold cyan]Syncing '
+                f'[green]{source_message}[/] to '
+                f'[green]cos://{self.region}/{self.name}/[/]'):
+            data_utils.parallel_upload_rclone(
```

Comment:
Is there a reason why `data_utils.parallel_upload_rclone` is implemented instead of using `data_utils.parallel_upload`? `data_utils.parallel_upload_rclone` doesn't group the files with `_group_files_by_dirs`. This would cause file number of `rclone copy` commands to be called when subset of files in a directory is specified as a source. With `data_utils.parallel_upload`, we can reduce the number of calls to `rclone copy` to number of directories being synced instead of every single files when a subset of files in a directory is specified as a source.

Reply:
`_group_files_by_dirs` didn't fit with my former implementation, but i believe it's no longer the case, so I'm trying to drop `parallel_upload_rclone` altogether in favor of `parallel_upload`. Will update tomorrow.

Reply:
Update @landscapepainter:
Removed `parallel_upload_rclone`. Code adjusted to rely on `parallel_upload`.